### ISSUE TITLE
Docs: Add short introduction to Docker on ARM (Raspberry Pi)

### ISSUE THREAD

**Issue by alexellis** [moby/moby#25810]
I would like to put together a quick guide for Docker on ARM - especially Raspberry Pi. There is a `curl | sh` method newly available which makes the process quicker and easier. From my perspective this should feel as close as possible to the experience of Docker on x86_64. 

I'd like to put a short/simple guide together:
- Recommended (default) OS to be Raspbian (Jessie Lite)
- `curl | sh` recommended installation method
- A mention to the problem with multi-arch - i.e. `docker run busybox` won't work
- A recommended base image: `resin/rpi-raspbian`
- Give short example of Dockerfile for Node.js, Python
- Short mention of GPIO (accessing the input output headers on the device) and privilege mode.
- A disclaimer / buyer beware.. few words on limitations.

Potential follow-up guide:
- Building Docker on ARM
- Adding a swap file
- Expected build duration
- Limitations

Ref: 
https://github.com/docker/docker/issues/24561

Input especially welcome from @DieterReuter , @estesp , @crosbymichael @StefanScherer as follow-up. @thaJeztah what are your thoughts?

**Comment by alexellis** [comment]
Additional material:

https://github.com/alexellis/docker-arm/blob/master/ZERO.md

https://github.com/alexellis/docker-arm

**Comment by justincormack** [comment]
I think that would be great. I don't think raspbian needs to be the recommendation - looks like we have Debian Jessie and Ubuntu Trusty debs too, and no doubt more will be added, and some distros are shipping Docker on arm too, so best to just list what is available.

**Comment by alexellis** [comment]
I guess I need input on whether it's useful to have a guide for a Raspberry Pi - and to cover all bases that would need to use a distro which is compatible with ARMv6 as well as ARMv7 (excludes Ubuntu and possibly also Debian Jessie)

The more generic the instructions, the less verbose/helpful the text can be. The one thing that will be common for the three supported debs is being able to type in something like:

* Installation:

```
$ curl -sSL test.docker.com | sh
```

There is also the issue of what base image to suggest. @thaJeztah mentioned @tianon is maintaining a set of images under the user `armhf` but these appear to only work on an ARMv7 device and not an ARMv6 device such as the Pi Zero / B / B+. The image @DieterReuter uses for the Raspbian build uses a base image of `resin/rpi-raspbian` which is compatible with both armv6 and armv7.

**Comment by justincormack** [comment]
Oh I see, I wasnt sure of the focus when you said "ARM - especially Raspberry Pi". Yes neither Jessie nor Ubuntu run on all the Pis, though they do run on the later Armv7 ones. I think focusing on Pi and occasionally calling out differences for other arm platforms would be fine.

The base images are an issue, as the official ones are armhf so v7, although in practise I suspect a lot work on v6 too, as the differences are fairly small, but obviously it is annoying if some things break (eg I am fairly sure that the v7 busybox image will work fine). There is an Alpine Linux build specifically for pi, ie v6 armhf, I could make a base image from that too, and a busybox image might make sense.

**Comment by justincormack** [comment]
As far as I can see `alpine` for armhf is compiled for armv6, so it is fine for both as a base image. There is not an official one, but there are several unofficial ones eg I have one at `docker pull justincormack/armhf-alpine` (also includes qemu static so can run on x86).

**Comment by alexellis** [comment]
`justincormack/armhf-alpine` works on my Pi Zero 👍  - that would help people with experience of Alpine on x86 transition to ARM. It's also a magnificent 3mb in size, wow!

For people moving from the Raspberry Pi world, they will need something with `apt-get` or it will probably be a lot to ask - plus compatibility with GPIO and hardware is well established. Do you think there is room for introducing both images and then giving a couple of examples with either?

Would you be able to propose a skeleton for the docs page?

**Comment by justincormack** [comment]
Oh it turns out there is an official `armhf/alpine` that I missed. That should be even smaller as it does not have the qemu for enabling running on z86 as well...

I think covering both is good. People use a mix of Debian and alpine in x86 world too, depending on what they are used to and whether size matters for what they are doing.

**Comment by thaJeztah** [comment]
ping @sfsmithcha where do you think we should put this documentation?

**Comment by alexellis** [comment]
@justincormack  This might even be suitable for airplane wifi!

```
latest: Pulling from armhf/alpine
c00d567ee11d: Extracting 1.081 MB/1.985 MB
```

and.. it does work on v6 / PiZero.

**Comment by thaJeztah** [comment]
@alexellis looks like the [armhf/alpine](https://hub.docker.com/r/armhf/alpine/) hasn't been updated for two months, so handle with care :smile:

ping @tianon think it's ok to mention/use those images in the documentation?

**Comment by tianon** [comment]
Heh, https://github.com/docker-library/official-images/blob/master/library/alpine was last updated two months ago as well (the two are somewhat unrelated, but amusing that they line up nicely).

Yeah, those images should be reasonably OK to talk about in the docs. :+1:

**Comment by thaJeztah** [comment]
Oh, time flies :smile: thanks for looking @tianon - do you also know why other images there don't work on ARMv6? https://github.com/docker/docker/issues/25810#issuecomment-240559153 (sorry, I'm a complete n00b on the ARMvWhatever differences)

**Comment by tianon** [comment]
That'd be because Raspberry Pi decided to use ARMv6 with floating point instead of the more common ARMv7 with floating point (which is what Debian, Ubuntu, etc all support, and which is why "Raspbian" even exists -- they had to rebuild Debian's ARMv7 "armhf" for the ARMv6 platform).

**Comment by sfsmithcha** [comment]
@thaJeztah , I think this is the perfect type of content for /docker/labs

@ManoMarks can you chime in?

**Comment by thaJeztah** [comment]
@sfsmithcha possibly; although we may need some (minimal) instructions on how to install docker in the installation guide?

**Comment by StefanScherer** [comment]
Installing Docker with `docker-machine create -d generic` is partly working for ARM as well, some corrections for Raspbian are on the way https://github.com/docker/machine/pull/3605

**Comment by alexellis** [comment]
@sfsmithcha  are you suggesting no instructions at all for ARM/Pi on the official docs site? This would be a real shame, I think that's the point of this issue being raised. We have lots of good guides available from Hypriot and myself but nothing official - however brief or unbiased. Now that curl|sh works as an installation method for Debian-like OSes on ARM we wanted to have some more visibility.

**Comment by ManoMarks** [comment]
My feeling is we should have an installation guide here: https://docs.docker.com/engine/installation/linux/ and then definitely a lab on docker/labs for running on multiple devices

**Comment by sfsmithcha** [comment]
@ManoMarks, @alexellis,

First off : @alexellis -- this is amazing content and a really good tutorial, don't get me wrong. On the other hand, l believe it should not be considered a shame to have content featured on labs. Also, as far as I understand it labs is Docker official.

The docs team is working to get tighter integration with labs so that you wouldn't have to look in multiple places for this type of content, which is, unfortunately, currently the case.

But adding this to the core docker installation docs creates an additional maintenance issue in the docs for content that is not Docker-specific. If ARM/pi is an officially supported environment, we should have it listed as an officially supported environment. (This is a lack in the current docs). I'm of the opinion we need to have a better installation guide experience than a bunch of branching guides for every possible platform.

I'm not going to put effort into blocking this PR from core docs. It is definitely not my goal  to prevent it from being avialable. However, please realize that once it's in the docs it may be subject to any pruning we need to do to improve the overall docs experience.

cc: @johndmulhausen

**Comment by DieterReuter** [comment]
I think for an initial documentation it would be great and sufficient to have at least the basic installation steps and the tested/supported OSes for ARM. So we can make sure that we list successfully tested OSes only! I've done the complete tests so far for all the recent Raspbian releases with 1.12.1-rc1 and can start this list with already tested and supported Raspbian releases here.

| Raspbian image file | Tested | Supported |
|---------------------|--------|-----------|
| 2016-05-10-raspbian-jessie-lite.zip | OK | YES |
| 2016-05-10-raspbian-jessie.zip | OK | YES |
| 2016-05-27-raspbian-jessie-lite.zip | OK | YES |
| 2016-05-27-raspbian-jessie.zip | OK | YES |

Once the Docker Engine 1.12.1 is officially released, the installation procedure on a Raspbian system would be the following:

**MAIN** branch:
```
$ curl -sSL https://get.docker.com | sh
```

**EXPERIMENTAL** branch:
```
$ curl -sSL https://experimental.docker.com | sh
```

**TESTING** branch: (only contains -rc versions)
```
$ curl -sSL https://test.docker.com | sh
```

For Raspbian the Docker storage-driver will be set to `overlay` by default. This is done by creating a systemd drop-in file at `/etc/systemd/system/docker.service.d/overlay.conf`.

Additionally we could recommend to add user `pi` to group `docker` so the user can use docker commands without the need of sudo:
```
$ sudo usermod -aG docker pi
```
To activate this settings the user has to logout and login again.

**Comment by alexellis** [comment]
Is everyone OK with us mentioning the issue around multi-arch and suggesting 2-3 tried/tested images such as those mentioned by myself and @justincormack ? These will work on armv6 and armv7 devices.

* armhf/alpine
* resin/rpi-raspbian

I guess my concern is that if we don't make mention of the fact most Docker hub images will fail to run, then lots of issues could get raised?

@ManoMarks I don't mind adding to docker/labs. What I am trying to say is that I think there should be a mention on the official docs that this is a thing and then give some sort of documentation under https://docs.docker.com/engine/installation/linux/ ?

I didn't mean my blog post to be a lift + drop into the official docs, I wanted to put together a guide that showed some of the steps needed for a good experience of Docker on the Pi especially.

**Comment by ManoMarks** [comment]
@alexellis I agree that we should have info in official docs about both the availability of these three distros and the need to use arm base images.

**Comment by sfsmithcha** [comment]
@alexellis do you want to go ahead and do a PR per @ManoMarks suggestion? At the path `docs/engine/installation/linux`.

**Comment by johndmulhausen** [comment]
Hi guys. John, the lead tech writer chiming in (thx for the CC, Charles). I think we're saying the same thing:

- Go ahead and add a detailed installation page in labs. 
- We can give a shoutout to the ARM support at https://docs.docker.com/engine/installation/ and link to the page in labs.

But, until such time as ARM support is more official, I don't think the docs team should be documenting it. 

What Charles is saying about the direction of the docs is true. We want to surface more things and be a hub for exploring the container ecosystem. But we are never going to be a one-stop shop that has full docs for everything the community does and third parties do; that's far too large of a scope for our little team.

The TL;DR is if you don't mind maintaining it, we don't mind shouting out. :)

**Comment by ManoMarks** [comment]
Also, we can link off to the labs repo from the official docs, once we get something on docker/labs

**Comment by justincormack** [comment]
I think there is an official arm release of raspbian Jessie with 1.12.1...
Hence the requirement for docs.

**Comment by withinboredom** [comment]
It's in the release notes, but not in the apt repo the script installs...

**Comment by alexellis** [comment]
I wanted to bring this thread back.. my initial intention was to provide some light-weight documentation for ARM on Docker and to help people understand the differences / get started.

Here are a couple of resources I put together that people may see while Googling for `ARM / Raspberry Pi Docker Docs` etc

http://blog.alexellis.io/5-things-docker-rpi/
5 things about Docker on Raspberry Pi

http://blog.alexellis.io/getting-started-with-docker-on-raspberry-pi/
Get Started with Docker 1.12 on Raspberry Pi

@johndmulhausen do you think we can re-visit this? What's your feeling on documentation now?

**Comment by alexellis** [comment]
@icecrime referencing https://github.com/docker/docker/issues/11580

Pinging @johndmulhausen again.. previous ping was a month ago already.

**Comment by thaJeztah** [comment]
Tried pinging him on Slack :innocent:

**Comment by alexellis** [comment]
Did it work? :-)

**Comment by thaJeztah** [comment]
Apparently not -_-

**Comment by johndmulhausen** [comment]
Hey @alexellis - thanks for being patient. 

So, we have a generic "Debian or Raspbian" installation page at https://docs.docker.com/engine/installation/linux/debian/

Perhaps a good approach would be to break out Raspbian to a separate guide (at /engine/installation/linux/raspbian/), have it use the ARM bits in the instructions, and call out any extra steps for ensuring that rPI's can get up and running.

CC'ing @mstanleyjones with her thoughts on that approach; it's a little special case compared to what she has been trying to do with a more unified install guide w/reusable content. 

What do you think, @alexellis ?

**Comment by StefanScherer** [comment]
I like the generic doc at https://docs.docker.com/engine/installation/linux/debian/ 👍

**Comment by thaJeztah** [comment]
@StefanScherer @alexellis this still needed?

**Comment by StefanScherer** [comment]
@thaJeztah everything fine for me, thanks.

**Comment by thaJeztah** [comment]
Let's close this for now; if we still want more information, perhaps it's best to open a new issue in the documentation issue tracker (https://github.com/docker/docker.github.io/issues
